File: baekjoon/baekjoon_1937.cpp
#include <iostream>
#include <vector> 

using namespace std; 

struct IntPair{
	int x, y; 
	IntPair(int x, int y){
		this->x = x;
		this->y = y; 
	}
}; 

vector<IntPair> getPossiblePaths(int** bambooMap, int n, int x, int y){
	vector<IntPair> possiblePaths; 

	if(x-1 >= 0 && bambooMap[x-1][y] > bambooMap[x][y])
		possiblePaths.push_back(IntPair(x-1, y));

	if(x+1 <  n && bambooMap[x+1][y] > bambooMap[x][y])
		possiblePaths.push_back(IntPair(x+1, y));

	if(y-1 >= 0 && bambooMap[x][y-1] > bambooMap[x][y])
		possiblePaths.push_back(IntPair(x, y-1));

	if(y+1 <  n && bambooMap[x][y+1] > bambooMap[x][y])
		possiblePaths.push_back(IntPair(x, y+1));

	return possiblePaths;
}

void getMaxDays(int** bambooMap, int** maxDayMap, int n, int x, int y){
	vector<IntPair> possiblePaths = getPossiblePaths(bambooMap, n, x, y);

	// base case
	if(possiblePaths.empty()) {
		maxDayMap[x][y] = 1;
		return;
	}

	//recursive case
	vector<int> possibleDays; 

	for (IntPair coord : possiblePaths) {
		if(maxDayMap[coord.x][coord.y] == 0)
			getMaxDays(bambooMap, maxDayMap, n, coord.x, coord.y);
		possibleDays.push_back(maxDayMap[coord.x][coord.y]);
	}

	maxDayMap[x][y] = *max_element(begin(possibleDays), end(possibleDays)) + 1;
}

int main() {
	int n;
	cin >> n; 

	int** bambooMap = new int*[n]; 
	int** maxDayMap = new int*[n];

	for(int i = 0; i < n; i++){
		bambooMap[i] = new int[n];
		maxDayMap[i] = new int[n];
	}

	for(int i = 0; i < n; i++){
		for(int j = 0; j < n; j++){
			scanf("%d", &bambooMap[i][j]);
		}
	}

	for(int i = 0; i < n; i++){
		for(int j = 0; j < n; j++){
			if(maxDayMap[i][j] == 0) 
				getMaxDays(bambooMap, maxDayMap, n, i, j);
		}
	}

	int max = 0;

	for(int i = 0; i < n; i++){
		for(int j = 0; j < n; j++){
			if(maxDayMap[i][j] > max)
				max = maxDayMap[i][j];
		}
	}

	for(int i = 0; i < n; i++){
		delete bambooMap[i];
		delete maxDayMap[i];
	}
	delete[] bambooMap;
	delete[] maxDayMap;

	cout << max << endl;
}

File: baekjoon/baekjoon_1937.java
package com.company;

import java.util.Collections;
import java.util.Scanner;
import java.util.ArrayList;

class IntPair {
    public int x;
    public int y;
    public IntPair(int x, int y) {
        this.x=x;
        this.y=y;
    }
}

class MyMap {
    private int[][] map;
    public MyMap(int n) {
        map = new int[n][n];
    }
    public int getVal(int x, int y){
        return map[x][y];
    }
    public void setVal(int x, int y, int newVal) {
        map[x][y] = newVal;
    }
}

public class Main {

    private static ArrayList<IntPair> getPossiblePaths(MyMap bambooMap, int n, int x, int y){
        ArrayList<IntPair> possiblePaths = new ArrayList<>();

        if (x-1 >= 0 && bambooMap.getVal(x-1,y) > bambooMap.getVal(x,y))
            possiblePaths.add(new IntPair(x-1,y));
        if (x+1 <  n && bambooMap.getVal(x+1,y) > bambooMap.getVal(x,y))
            possiblePaths.add(new IntPair(x+1,y));
        if (y-1 >= 0 && bambooMap.getVal(x,y-1) > bambooMap.getVal(x,y))
            possiblePaths.add(new IntPair(x, y-1));
        if (y+1 <  n && bambooMap.getVal(x,y+1) > bambooMap.getVal(x,y))
            possiblePaths.add(new IntPair(x, y+1));

        return possiblePaths;
    }

    private static void getMaxDays(MyMap bambooMap, MyMap maxDayMap, int n, int x, int y){
        ArrayList<IntPair> possiblePaths = getPossiblePaths(bambooMap, n, x, y);

        // base case
        if (possiblePaths.size() == 0){
            maxDayMap.setVal(x, y, 1);
            return ;
        }

        ArrayList<Integer> possibleDays = new ArrayList();

        // recursive case
        for(IntPair coord : possiblePaths) {
            if(maxDayMap.getVal(coord.x, coord.y) == 0)
                getMaxDays(bambooMap, maxDayMap, n, coord.x, coord.y);
            possibleDays.add(maxDayMap.getVal(coord.x, coord.y));
        }

        maxDayMap.setVal(x,y, Collections.max(possibleDays) + 1);

        return;
    }

    public static void main(String[] args) {
        Scanner scan = new Scanner(System.in);

        int n = scan.nextInt();
        MyMap bambooMap = new MyMap(n);
        MyMap maxDayMap = new MyMap(n);

        for(int i = 0; i < n; i++){
            for(int j = 0; j < n; j++){
                bambooMap.setVal(i, j, scan.nextInt());
            }
        }

        for(int i = 0; i < n; i++){
            for(int j = 0; j < n; j++){
                if (maxDayMap.getVal(i,j) == 0){
                    getMaxDays(bambooMap, maxDayMap, n, i, j);
                }
            }
        }

        int max = 0;

        for(int i = 0; i < n; i++){
            for(int j = 0; j < n; j++){
                if(maxDayMap.getVal(i,j) > max)
                    max = maxDayMap.getVal(i,j);
            }
        }

        System.out.println(max);
    }
}
